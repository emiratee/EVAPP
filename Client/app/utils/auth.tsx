import React, { createContext, useContext, useEffect, useState } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';

const AuthContext = createContext();

export const useAuth = () => {
    return useContext(AuthContext);
};

export const AuthProvider = ({ children }) => {
    const [token, setToken] = useState(null);

    const login = (jwtToken) => {
         AsyncStorage.setItem('jwtToken', jwtToken)
            .then(() => {
                setToken(jwtToken);
            })
            .catch((error) => {
                console.error('Error storing JWT token:', error);
            });
    };

    const logout = () => {
        // Remove the JWT token from AsyncStorage
        AsyncStorage.removeItem('jwtToken')
            .then(() => {
                setToken(null);
            })
            .catch((error) => {
                console.error('Error removing JWT token:', error);
            });
    };

    const getData = async () => {
        try {
          const value = await AsyncStorage.getItem('jwtToken');
          if (value !== null) {
            // value previously stored
            setToken(value);
          }
        } catch (e) {
          console.log(e)
        }
      };
    
    useEffect(() => {
        // logout()
        getData();
        
    }, [])
    

    const isAuthenticated = !!token;

    const value = {
        token,
        setToken,
        login,
        logout,
        isAuthenticated,
    };

    return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};
