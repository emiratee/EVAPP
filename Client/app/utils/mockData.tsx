import { createContext, useContext, useEffect, useState } from 'react';
import React from 'react'
import * as Types from '../../types/types'

type MockDataType = {
    mockUsers: Types.TUser[];
    setMockUsers: React.Dispatch<React.SetStateAction<Types.TUser[]>>;
    trips: Types.TTrip[];
    setTrips: React.Dispatch<React.SetStateAction<Types.TTrip[]>>;
}


const defaultMockDataContext: MockDataType = {

    setMockUsers: () => { },
    mockUsers: [],
    setTrips: () => { },
    trips: []
};

const MockDataContext = createContext(defaultMockDataContext);

export const MockDataProvider = ({ children }: { children: React.ReactNode }) => {
    const [mockUsers, setMockUsers] = useState<Types.TUser[]>([
        {
            id: '234',
            name: 'Erik L.',
            memberSince: '23rd, Feb. 2023',
            email: 'erik@gmail.com',
            phoneNumber: '35326234112',
            password: '123456',
            cars: [
                {
                    id: "123",
                    model: 'Audi A4',
                    color: 'Black',
                    seats: 4,
                    licencePlate: 'SO-SI-6969'
                },
                {
                    id: "1234",
                    model: 'Audi A5',
                    color: 'Red',
                    seats: 2,
                    licencePlate: 'XXX-6969'
                },
            ],
            passengerRating: {
                totalReviews: 10,
                totalRating: 49,
                averageRating: 4.9
            },
            driverRating: {
                totalReviews: 5,
                totalRating: 10,
                averageRating: 2
            },
            tripsAsDriverIDs: ["123"],
            tripsAsPasangerIDs: [],
            credits: '20.00'
        }
    ])
    const [trips, setTrips] = useState<Types.TTrip[]>([
        {
            id: "123",
            departure: {
                country: "Germany",
                city: "Berlin",
                address: "Stresemannstraße 123, 10963 Berlin, Germany",
                time: "12:30",
            },
            destination: {
                country: "Germany",
                address: "C/ d'Àvila, 27, 08005 Barcelona, Spain",
                city: "Barcelona",
                time: "15:30",
            },
            date: "2023-11-28",
            totalTime: "3:00",

            seats: {
                available: 2,
                total: 5,
            },
            services: {
                alcohol: false,
                childSeat: false,
                luggage: true,
                pets: true,
                smoking: true,
                comments: "here is a comment",
            },
            selectedCar: {
                id: "123",
                model: 'Audi A4',
                color: 'Black',
                seats: 4,
                licencePlate: 'SO-SI-6969'
            },
            price: "10.25",
            driverID: "234",
            passengersIDs: [],
            succesful: false
        }

    ])




    return (
        <MockDataContext.Provider value={{ setMockUsers, mockUsers, setTrips, trips }} >
            {children}
        </MockDataContext.Provider>
    );
};

export const useMockData = () => {
    return useContext(MockDataContext);
}