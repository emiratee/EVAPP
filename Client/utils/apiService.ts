import { Alert } from "react-native";

// const BASE_URL = process.env.ATLAR_URL || 'https://evap-pserver-r1s4.vercel.app';
const BASE_URL = process.env.ATLAR_URL || 'http://127.0.0.1:3000'; //not working
const checkResponse = (response: Response): void => {
    if (!response.ok) throw new Error(`Request failed with status ${response.status}`);
};

const getFilteredTrips = async (departure, destination, date, seats): Promise<any> => {
    try {
        if (!destination) {
            const response = await fetch(`${BASE_URL}/trips?departureCountry=${departure.country}&departureCity=${departure.city}&date=${date}&seats=${seats}`);
            return await response.json();

        } else {

            const response = await fetch(`${BASE_URL}/trips?departureCountry=${departure.country}&departureCity=${departure.city}&destinationCountry=${destination?.country}&destinationCity=${destination?.city}&date=${date}&seats=${seats}`);
            return await response.json();
        }

    } catch (error) {
        console.error(error);
        throw error;
    }
}

const getDriver = async (driverId:string): Promise<any> => {
    try {
        const response = await fetch(`${BASE_URL}/user/${driverId}`);
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}

const getUser = async (token: string) => {
    try {
        const response = await fetch(`${BASE_URL}/user`, {
            method: 'GET',
            headers: {
                'Authorization': `${token}`
            }
        });
        checkResponse(response);
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}

const addCar = async (data, token: string) => {
    try {
        const response = await fetch(`${BASE_URL}/user/cars`, {
            method: 'PUT',
            headers: {
                "Content-Type": "application/json",
                'Authorization': `${token}`
            },
            body: JSON.stringify(data),
        });
        checkResponse(response);
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}

const addNewTrip = async (data, token: string) => {
    try {
        const response = await fetch(`${BASE_URL}/trip/create`, {
            method: 'POST',
            headers: {
                "Content-Type": "application/json",
                'Authorization': `${token}`
            },
            body: JSON.stringify(data),
        });
        checkResponse(response);
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}

const putTripsAsDriver = async (data, token: string) => {
    try {
        const response = await fetch(`${BASE_URL}/user/trips/driver`, {
            method: 'PUT',
            headers: {
                "Content-Type": "application/json",
                'Authorization': `${token}`
            },
            body: JSON.stringify(data),
        });
        checkResponse(response);
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}

const postRegister = async (data) => {
    try {
        const response = await fetch(`${BASE_URL}/user/account/register`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(data)
        })
        checkResponse(response);
        return await response.json();

    } catch (error) {
        Alert.alert('here')
        console.log(error)
        throw error;
    }
}

const postLogin = async (email: string, password: string) => {
    try {
        const response = await fetch(`${BASE_URL}/user/account/login`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ email, password })
        })
        //checkResponse(response);
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}

const getHistory = async (token: string) => {
    try {
        const response = await fetch(`${BASE_URL}/user/history`, {
            method: 'GET',
            headers: {
                'Authorization': `${token}`
            }
        });
        checkResponse(response);
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}

const putAvailableCredits = async (amount: String, token: String) => {
    try {
        const response = await fetch(`${BASE_URL}/user/credits/available`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `${token}`
            },
            body: JSON.stringify({ amount })
        })
        //checkResponse(response);
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}

const putApproveTrip = async (data, token: string) => {
    try {
        const response = await fetch(`${BASE_URL}/user/trips/approve`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `${token}`
            },
            body: JSON.stringify({ data })
        })
        //checkResponse(response);
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}
const putRejectTrip = async (data, token: string) => {
    try {
        const response = await fetch(`${BASE_URL}/user/trips/reject`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `${token}`
            },
            body: JSON.stringify({ data })
        })
        //checkResponse(response);
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}

const putRequestTrip = async (data, token: string) => {
    try {
        const response = await fetch(`${BASE_URL}/trips/request`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `${token}`
            },
            body: JSON.stringify({ data })
        })
        //checkResponse(response);
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}



const updateAccount = async (data, token: string) => {
    try {
        const { currentPassword, newPassword, image} = data;
        const response = await fetch(`${BASE_URL}/user/account/update`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `${token}`
            },
            body: JSON.stringify({ 
                currentPassword,
                newPassword,
                image
             })
        })
        return await response.json();
    } catch (error) {
        console.error(error);
        throw error;
    }
}

const cloudinaryUpload = async (data) => {
    try {
        const response = await fetch("https://api.cloudinary.com/v1_1/dzfbxhrtf/upload", {
            method: "POST",
            body: data,
        });

        if (!response.ok) {
            throw new Error('Network response was not ok');
        }

        const cloudinaryData = await response.json();
        
        return cloudinaryData.secure_url;
    } catch (err) {
        console.error('An Error Occurred While Uploading:', err);
        Alert.alert('An Error Occurred While Uploading');
        throw err;
    }
}

export {
    cloudinaryUpload,
    getFilteredTrips,
    getDriver,
    getUser,
    addCar,
    addNewTrip,
    putTripsAsDriver,
    postRegister,
    postLogin,
    putAvailableCredits,
    getHistory,
    putApproveTrip,
    putRejectTrip,
    putRequestTrip,
    updateAccount
}