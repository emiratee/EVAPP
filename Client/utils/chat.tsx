import React, { createContext, useContext, useState } from 'react';

const ChatContext = createContext();

export const useChat = () => {
    return useContext(ChatContext);
};

export const ChatProvider = ({ children }: { children: React.ReactNode }) => {
    const [name, setName] = useState('HOLA');
    const [imageUrl, setImageUrl] = useState();

    return <ChatContext.Provider value={{
        name,
        setName,
        imageUrl,
        setImageUrl
    }}>
        {children}
    </ChatContext.Provider>;
};