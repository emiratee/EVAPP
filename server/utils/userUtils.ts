import jwt from 'jsonwebtoken';
import path from 'path'
import dotenv from "dotenv";
import User from '../models/User';

dotenv.config({ path: path.join(__dirname, '..', '..', '.env') });

export async function validateUser(req: Request): Promise<any> {
    try {
        const { authorization } = req.headers;

        if (!authorization) return false;

        const userId = tokenToUserId(authorization);
        if (!userId) return false;

        const user = await User.findOne({ userId });

        if (!user) return false;

        return { userId, user };
    } catch (error) {
        throw error;
    }
}

export function tokenToUserId(token: string) {
    const SECRET_KEY = process.env.SECRET_KEY!;
    try {
        const decodedToken = jwt.verify(token, SECRET_KEY) as { userId: string };
        return decodedToken.userId;
    } catch (error) {
        return undefined;
    }
}