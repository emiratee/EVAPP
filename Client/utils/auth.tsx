import React, { createContext, useContext, useEffect, useState } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { getUser } from './apiService';
import { TUser, TAuth } from '../types/types';

type AuthContextType = {
    token: string | undefined;
    setToken: React.Dispatch<React.SetStateAction<string | undefined>>;
    user: types.TUser | null;
    setUser: React.Dispatch<React.SetStateAction<types.TUser | null>>;
    login: (token: string) => Promise<void>;
    logout: () => Promise<void>;
    isAuthenticated: boolean;
}

const defaultAuthContext: AuthContextType = {
    token: undefined,
    setToken: () => { },
    user: null,
    setUser: () => { },
    login: async () => { },
    logout: async () => { },
    isAuthenticated: false,
};

const AuthContext = createContext(defaultAuthContext);

export const useAuth = () => {
    return useContext(AuthContext);
};

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
    const [token, setToken] = useState<string | undefined>(undefined);
    const [user, setUser] = useState<types.TUser | null>(null);

    useEffect(() => {
        (async () => {
            if (token) {
                const data = await getUser(token);
                setUser(data);
            }
        })();
    }, [token])

    const login = async (token: string): Promise<void> => {
        try {
            await AsyncStorage.setItem('token', token)
            setToken(token);
        } catch (error) {
            console.error('Error storing token:', error);
        }
    };

    const logout = async (): Promise<void> => {
        try {
            await AsyncStorage.removeItem('token')
            setToken(null);
            setUser(null);
        } catch (error) {
            console.error('Error removing token:', error)
        }
    };

    const getData = async (): Promise<void> => {
        try {
            const value = await AsyncStorage.getItem('token');
            if (value !== null) setToken(value);
        } catch (error) {
            console.log(error)
        }
    };

    useEffect(() => {
        getData();
    }, []);


    const isAuthenticated: boolean = !!token;

    return <AuthContext.Provider value={{
        token,
        setToken,
        login,
        logout,
        isAuthenticated,
        user,
        setUser
    }}>
        {children}
    </AuthContext.Provider>;
};
