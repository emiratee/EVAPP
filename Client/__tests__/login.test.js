import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react-native';
import Login from '../app/(tabs)/login';
import { postLogin } from '../utils/apiService';


// Mock useAuth
// jest.mock('../utils/auth', () => ({
//     useAuth: () => ({
//         login: jest.fn(),
//         isAuthenticated: false, // Modify this to match your useAuth implementation
//     }),
// }));


// Mock postLogin
jest.mock('../utils/apiService', () => ({
    postLogin: jest.fn(() => ({ token: 'mockToken' })), // Modify this mock response as needed
}));



describe('Login Page', () => {
    it('renders correctly', () => {
        const tree = renderer.create(<Login />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('handles form submission with valid data', async () => {
        const { getByPlaceholderText, getByText } = render(<Login />);
        const emailInput = getByPlaceholderText('E-mail');
        const passwordInput = getByPlaceholderText('Password');
        const loginButton = getByText('Login');

        // Simulate user input
        fireEvent.changeText(emailInput, 'test@example.com');
        fireEvent.changeText(passwordInput, 'password');

        // Mock the postLogin function
        // postLogin.mockResolvedValueOnce({ token: 'mockToken' }); // Use a mock response

        // Simulate button click
        fireEvent.press(loginButton);

        // Assert that postLogin function is called with correct arguments
        expect(postLogin).toHaveBeenCalledWith('test@example.com', 'password');
    });

    it('displays error messages for empty inputs', async () => {
        const { getByText } = render(<Login />);
        const loginButton = getByText('Login');

        // Simulate button click with empty inputs
        fireEvent.press(loginButton);

        // Assert that error messages are displayed
        expect(getByText('Please enter e-mail')).toBeTruthy();
        expect(getByText('Please enter password')).toBeTruthy();
    });

});
