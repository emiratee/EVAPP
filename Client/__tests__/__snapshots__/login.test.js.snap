// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "flex": 1,
      "justifyContent": "center",
      "padding": 30,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#ededed",
        "borderRadius": 10,
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "marginBottom": 15,
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      collapsable={false}
      handlerTag={1}
      handlerType="NativeViewGestureHandler"
      keyboardType="email-address"
      onChangeText={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      placeholder="E-mail"
      placeholderTextColor="#838383"
      style={
        [
          {
            "backgroundColor": "#ededed",
            "borderRadius": 10,
            "flex": 1,
            "fontSize": 16,
            "height": 50,
            "marginBottom": 15,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
            "top": 8,
            "zIndex": 1,
          },
          false,
        ]
      }
      value=""
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#ededed",
        "borderRadius": 10,
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "marginBottom": 15,
      }
    }
  >
    <TextInput
      collapsable={false}
      handlerTag={2}
      handlerType="NativeViewGestureHandler"
      onChangeText={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      placeholder="Password"
      placeholderTextColor="#838383"
      secureTextEntry={true}
      style={
        [
          {
            "backgroundColor": "#ededed",
            "borderRadius": 10,
            "flex": 1,
            "fontSize": 16,
            "height": 50,
            "marginBottom": 15,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
            "top": 8,
            "zIndex": 1,
          },
          false,
        ]
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#000",
        "borderRadius": 25,
        "marginTop": 10,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "#fff",
        }
      }
    >
      Login
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "marginTop": 10,
        "opacity": 1,
      }
    }
  />
</View>
`;
